### PR TITLE
Refactors and adds capability to createtransfers

### LINKED ISSUE #16: Add more information to stderr and/or exit status in createtransfers.py
We could do with a little more information being output to the command line to get a feel for whether writing the files has worked at all. Plus any other relevant output we/a user might need.

### LINKED ISSUE #24: Problem: missing cohesive build method
The AM Ansible role [has been recently updated](https://github.com/artefactual-labs/ansible-archivematica-src/pull/172) so it runs [`createtransfers.py`](https://github.com/artefactual/archivematica-sampledata/blob/9805b0a0bf1dad65236ffa36b4db0b79b709ae74/createtransfers.py) once the repo is downloaded. I think that we should introduce a `Makefile` in this repo so the build process details are hidden from the consumers. This would allow us greater flexibility in controlling what's in the build.

### LINKED ISSUE #25: Issue: Need the capability to create large transfers e.g. 1000+ files, 200+ folders
As a tester I need to be able to create fairly arbitrarily sized transfers to be able to test Archivematica's limits. An issue where the limits of AM were noticed, and thus fixed is [here](https://github.com/artefactual/archivematica/issues/821#issue-278984149). 

Have a look at creating this sample set in the create transfers utility for testing purposes.

### PR DESCRIPTION
**Pull request by ross-spencer**
The most substantial change of this refactor which affects the code review somewhat is moving the script down a level to its own module level folder. This has been recognised by Git as a deletion of the original code. (I have made various attempts to recover a diff-ed version but currently the results are to no avail) 

`createtransfers.py` has been modified to do the following:

* Adds create-deep-transfers functionality
* Refactors args to be more consistent with other Artefactual exes
* Functionality of new args refactoring should be equivalent to previous tool
* Adds a unit test to help us to provide users feedback on create-deep-transfer
* Adds logging to all functions where it is appropriate to provide feedback
* Adds functionality to handle tuples for multiple filename generation
* Adds emoji as a curio for users to Archivematica to test against
* Returns non-zero where operations fail
* Complies with Flake8/PEP8
* Resolves #16
* Resolves #25 

Submitted in support of testing: https://github.com/artefactual/archivematica/issues/821#issue-278984149 and future similar cases.

### CODE REVIEW

File: createtransfers/createtransfers.py [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+    if not os.path.exists(file_path):
+        logger.error("Cannot find media file at location %s", file_path)
+        return ERR_OPERATION_FAILED
+    logger.info(("%s duplicated %s times "
+                 "will take up %sGB disk space"), file_path, n,
+                ((os.path.getsize(file_path)
+                 * n) / 1024.0) / 1024.0 / 1024.0)
+
+
+def create_variously_encoded_files(**kwargs):
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: Unused `kwargs` here are not needed.

File: createtransfers/createtransfers.py [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+    video_path = os.path.join(
+        HERE, 'SampleTransfers', 'Multimedia', 'MakeUp.mov')
+
+    success = create_transfer_collection(performance_path,
+                                         video_path,
+                                         video_n,
+                                         'video-14M-each',
+                                         'mov')
+    if success is not None:
+        ERR_OPERATION_FAILED
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: Shouldn't this return `ERR_OPERATION_FAILED`? That said, given my other comment, the code should be changed to use exceptions instead of returning `success`/`fail` type values.

File: createtransfers/createtransfers.py [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+                total_dirs, total_files)
+
+    # We return this structure as an output of the unit testing developed
+    # to understand the structure that we should output here.
+    return FolderStructure(depth=structure.depth,
+                           dirs=total_dirs,
+                           files=total_files)
+
+
+def create_folders(folder_list, target, structure):
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: Unused argument `target`.

(We should add tox and flake8 or pylint to this repo so that these kind of linter issues can be caught without a human.)

File: createtransfers/createtransfers.py [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+
+from __future__ import division
+
+from collections import namedtuple, OrderedDict
+import os
+import sys
+import shutil
+
+import createtransfersargs
+import loggingconfig
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py:29]: My linter (pylint) says the following:

    W: 26, 0: Relative import 'createtransfersargs', should be 'createtransfers.createtransfersargs' (relative-import)
    W: 27, 0: Relative import 'loggingconfig', should be 'createtransfers.loggingconfig' (relative-import)

File: createtransfers/createtransfers.py [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+        'encoding': 'utf-8',
+        'file_name': (u'hearts-{0}.txt'.format(
+                      u'\u2764\uD83D\uDC96\uD83D\uDC99'
+                      u'\uD83D\uDC9A\uD83D\uDC9B\uD83D'
+                      u'\uDC9C\uD83D\uDC9D'),
+                      u'chess-{0}.txt'.format(
+                      u'\u2655\u2656\u2657\u2658\u2659'
+                      u'\u265A\u265B\u265C\u265D\u265E'
+                      u'\u265F'))
+    }
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: Nitpick: indentation could be improved here for readability, suggesting:

    ...
        'file_name': (
            u'hearts-{0}.txt'.format(
                u'\u2764\uD83D\uDC96\uD83D\uDC99'
                u'\uD83D\uDC9A\uD83D\uDC9B\uD83D'
                u'\uDC9C\uD83D\uDC9D'),
            u'chess-{0}.txt'.format(
                u'\u2655\u2656\u2657\u2658\u2659'
                u'\u265A\u265B\u265C\u265D\u265E'
                u'\u265F'))

File: createtransfers/createtransfers.py [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+    file_path_bytes = os.path.join(file_path, file_name_bytes)
+    try:
+        with open(file_path_bytes, 'w') as fout:
+            fout.write('The name of this file is encoded with encoding'
+                       ' {}\n'.format(encoding))
+            logger.info("Created: %s", file_path_bytes)
+    except OSError as e:
+        logger.error("Failed to open file: %s, error: %s",
+                     file_path_bytes, e)
+        return ERR_OPERATION_FAILED
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: I think that this entire pattern of catching errors and returning error codes is unpythonic and should be changed. See https://eli.thegreenplace.net/2008/08/21/robust-exception-handling/. We should be raising custom exceptions here if necessary and capturing them when things go wrong, not checking for an error code in the return value of a func that we expect to return `None`.

File: createtransfers/createtransfers.py [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+    image_n = 113
+    video_n = 669
+
+    performance_path = os.path.join(
+        HERE, 'TestTransfers', 'acceptance-tests', 'performance')
+    rm_dirs_and_create(performance_path)
+
+    # Create dir images-17M-each-2G-total/ containing 2G of 17M image files.
+    image_path = os.path.join(
+        HERE, 'TestTransfers', 'manualNormalization', 'manualNormalization',
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: Please restore the capitalization to `'ManualNormalization'`

File: createtransfers/createtransfers.py [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+            success = rm_dirs_and_create(path)
+            if success is not None:
+                return success
+            for i in range(structure.files):
+                new_path = os.path.join(path,
+                                        create_folder_string(i, "md"))
+                shutil.copyfile(default_file, new_path)
+
+
+def create_deep_transfers(**kwargs):
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: If `recursion_depth` etc are required args, why not just specify them as params to the func instead of using `**kwargs`. Seems more in line with "Explicit is better than implicit." Note you can define `def create_deep_transfers(create_deep_transfers, number_of_directories, number_of_files):` and then still call it with `create_deep_transfers(**{'create_deep_transfers': ...})`.

File: createtransfers/createtransfers.py [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+    """ Create a default object. Primarily for use with the deep-transfers
+    functionality of this tool, but could equally be used elsewhere.
+    """
+    rm_dirs_and_create(target_path)
+
+    default_file_path = os.path.join(target_path, "README.md")
+
+    default_string = ("# Archivematica Deep Transfer Sample Set\n\n"
+                      "An arbitrarily generated depth of folders and files\n"
+                      "to test the recusive limits of Archvematica and its\n"
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: Typo: "recusive".

File: createtransfers/createtransfers.py [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+                      "capability to generate metadata.\n")
+
+    logger.info("Creating default file to copy into deep "
+                "transfer locations, %s", default_file_path)
+
+    f = open(default_file_path, 'w')
+    f.write(default_string)
+    f.close()
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: Use the `with` context manager here to ensure `f` gets closed even if an exception is raised.

File: createtransfers/createtransfers.py [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+
+    return default_file_path
+
+
+def create_folder_string(folder_no, ext=None):
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: It seems odd that this func has "folder" in the name when clearly it deals with extensions and filenames as well as folder names. Can it be renamed so that it's clear it is used to build directory AND file names?

File: createtransfers/createtransfers.py [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+                           dirs=total_dirs,
+                           files=total_files)
+
+
+def create_folders(folder_list, target, structure):
+    new_folder_list = []
+    for f in folder_list:
+        for folder_no in range(structure.dirs):
+            new_folder_list.append(os.path.join(f,
+                                   create_folder_string(folder_no)))
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: Nitpick: The indentation is off here. In general, the args to a func should be at the same indentation level. As it stands, it makes it look like `os.path.join...` and `create_folder_string` are both args to `append(`.

File: createtransfers/createtransfersargs.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        if subcommand.args is not None:
+            for arg in subcommand.args:
+                subparser.add_argument(
+                    arg.name, help=arg.help, type=arg.type)
+
+        if subcommand.opts is not None:
+            # Opts is a tuple, but subcommand.opts is a tuple as well.
+            # If we have the former, we don't want to iterate, we only have a
+            # single instance. If we have the latter, we want to iterate
+            # and create a new set of optional arguments per Opt item.
```
Comment (jrwdunham) [inline createtransfers/createtransfersargs.py]: This seems needlessly complicated. Can't we just always assume that `subcommand.opts` if not `None` is *always* going to be a tuple of namedtuples? If so, we can just use lines 94-100 and remove the rest.

File: createtransfers/createtransfers.py [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+def create_deep_transfers(**kwargs):
+
+    folder_structure = FolderStructure(depth=kwargs["recursion_depth"],
+                                       dirs=kwargs["number_of_directories"],
+                                       files=kwargs["number_of_files"])
+
+    # create a default folder from which to work from clear of the rest of
+    # our environment.
+    target_path = os.path.join(
+        HERE, 'TestTransfers', 'deep_transfer')
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py:366]: I think you should add this to .gitignore so it doesn't get committed accidentally in the future, i.e., add this line to .gitignore:

    TestTransfers/deep_transfer/

File: createtransfers/createtransfers.py (hunk 1 above)
Comment (jrwdunham) [reply createtransfers/createtransfers.py]: Having read to the bottom, I see why you've made all subcommand handlers take `**kwargs`. However, I would prefer it if we could not have unused params in our func definitions. Could we add some configuration to `COMMANDS` below so that `create_deep_transfers` gets called with kwargs but the others do not?

**Review by jrwdunham — CHANGES_REQUESTED** [review]
I have tested the new subcommand introduced here and can confirm it works. The pre-existing subcommands continue to work also.

I have some comments and requests for changes. The biggest request is that the pattern of returning error codes be replaced with a more Pythonic pattern of raising appropriate exceptions.

File: createtransfers/createtransfers.py (hunk 7 above)
Comment (ross-spencer) [reply createtransfers/createtransfers.py]: @jrwdunham can you clear your sampledata repository here and see if the folder still exists. In my working copy there wasn't such a folder so this change was to make the generation of this set work.

File: createtransfers/createtransfers.py (hunk 11 above)
Comment (ross-spencer) [reply createtransfers/createtransfers.py]: :+1:

File: createtransfers/createtransfers.py (hunk 14 above)
Comment (ross-spencer) [reply createtransfers/createtransfers.py:366]: :+1:

File: createtransfers/createtransfers.py (hunk 5 above)
Comment (ross-spencer) [reply createtransfers/createtransfers.py]: :+1:

File: createtransfers/createtransfersargs.py (hunk 13 above)
Comment (ross-spencer) [reply createtransfers/createtransfersargs.py]: Hmm, that does work. I was probably just overthinking that.

File: createtransfers/createtransfers.py (hunk 4 above)
Comment (ross-spencer) [reply createtransfers/createtransfers.py:29]: Removing __init__.py fixed this.

**Comment by ross-spencer** [comment]
Hi @jrwdunham I've made those changes over three commits. The robust exception handling commit should be the cleanest, and the one i'm expecting I'll need to look at again. But let me know. 

Have you any decent resources on adding tox to a project? I haven't done that before. I think perhaps we can open up a new ticket for that. I'll be happy to take care of that though once it is open.

File: createtransfers/createtransfers.py [hunk 15]
```diff
@@ -0,0 +1,413 @@
...
+                       ' {}\n'.format(encoding))
+            LOGGER.info("Created: %s", file_path_bytes)
+    except (OSError, IOError) as err:
+        ret = "Failed to open file: {0}, error: {1}".format(file_path_bytes,
+                                                            err)
+        raise CreateTransferException(ret)
+
+
+def rm_dirs_and_create(dir_path):
+    """Remove pre-existing director structures to write new ones out
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: Typo: "director"

File: createtransfers/createtransfers.py [hunk 16]
```diff
@@ -0,0 +1,413 @@
...
+    """
+    if os.path.exists(dir_path):
+        try:
+            shutil.rmtree(dir_path)
+        except OSError as err:
+            ret = "Failed to remove dirtree: {0} {1}".format(dir_path, err)
+            raise CreateTransferException(ret)
+    try:
+        os.makedirs(dir_path)
+    except (os.error) as err:
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: Is there a reason to use `os.error` here when you're using `OSError` above? Can we just use `OSError` to be consistent? Cf. https://stackoverflow.com/questions/989259/whats-the-point-of-os-error#comment797961_989266.

File: createtransfers/createtransfers.py (hunk 9 above)
Comment (jrwdunham) [reply createtransfers/createtransfers.py]: Typo still present.

File: createtransfers/createtransfers.py [hunk 17]
```diff
@@ -0,0 +1,413 @@
...
+    """To create our transfers we are copying from elsewhere within the
+    sample data repository to generate a new collection.
+    """
+
+    # Stat file, and return asap if it doesn't exist. If it does exist,
+    # provide the user an idea about the size of the collection being
+    # generated.
+    try:
+        stat_file(media_path, media_n)
+        new_media_dir = os.path.join(collection_path, new_media_dirname)
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: No need for this line to be in the scope of the `try:`

File: createtransfers/createtransfers.py [hunk 18]
```diff
@@ -0,0 +1,413 @@
...
+                    LOGGER.error(err)
+                    return
+        else:
+            try:
+                create_file_and_write(encoding_info['file_name'],
+                                      encoding_dir_path,
+                                      encoding)
+            except CreateTransferException as err:
+                LOGGER.error(err)
+                return
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py:190]: Lines 177 to 193 could be replaced with this:

        file_names = encoding_info['file_name']
        if not isinstance(file_names, (tuple, list)):
            file_names = (file_names,)
        for fname in file_names:
            try:
                create_file_and_write(fname,
                                      encoding_dir_path,
                                      encoding)
            except CreateTransferException as err:
                LOGGER.error(err)
                return

File: createtransfers/createtransfers.py (hunk 7 above)
Comment (jrwdunham) [reply createtransfers/createtransfers.py]: I did a fresh clone of this repo and the capitalization is `archivematica-sampledata/TestTransfers/ManualNormalization/manualNormalization/`. See https://github.com/artefactual/archivematica-sampledata/tree/master/TestTransfers/ManualNormalization/manualNormalization. I'm a bit confused here ...

File: createtransfers/createtransfersargs.py [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+        name='number_of_files',
+        metavar='INT',
+        help='Number of files per sub-directory. '
+             'Default {0} files.'.format(DEFAULT_NUMBER_FILES),
+        default=DEFAULT_NUMBER_FILES,
+        type=int)
+
+    # Sub-command configuration: give them a name, help text, a tuple of
+    # ``Arg`` instances and a tuple of ``Opts`` instances.
+    subcommand = namedtuple('SubCommand', ['name', 'help', 'args', 'opts'])
```
Comment (jrwdunham) [inline createtransfers/createtransfersargs.py]: I think your initial spelling for this namedtuple (`SubCommand`) was correct. Why did you change it to `subcommand`? It's essentially a class so it should follow the Python class naming conventions. Now it looks like you're calling a function.

File: createtransfers/createtransfers.py [hunk 20]
```diff
@@ -0,0 +1,413 @@
...
+
+def main():
+    """Entrypoint for createtransfers.py"""
+    cmd_deep = 'create-deep-transfers'
+    cmd_large = 'create-large-test-transfers'
+    cmd_encodings = 'create-variously-encoded-files'
+
+    # Create a named tuple to help us control what is passed
+    # to each of the various commands.
+    command = namedtuple('Command', 'cmd_name use_kwargs')
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: A var referencing the return value of `namedtuple()` should follow "class factory function" naming conventions, i.e., `command` should be `Command` here. Cf. https://stackoverflow.com/a/36371122/992730.

File: createtransfers/createtransfersargs.py [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+DEFAULT_DEPTH = 5
+DEFAULT_NUMBER_DIRS = 3
+DEFAULT_NUMBER_FILES = 4
+
+
+def get_parser(**kwargs):
+    """Argument parsing configuration for the CLI."""
+
+    # Reusable option constants (for CLI).
+    opt = namedtuple('Opt', ['name', 'metavar', 'help', 'default', 'type'])
```
Comment (jrwdunham) [inline createtransfers/createtransfersargs.py]: `opt` should be `Opt` here. See related comments for justification.

File: createtransfers/createtransfersargs.py [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+    depth = opt(
+        name='recursion_depth',
```
Comment (jrwdunham) [inline createtransfers/createtransfersargs.py]: The `name` values of these `Opt` instances should be hyphen-cased, e.g., `'recursion-depth'`. That way, the CLI will allow for the more natural `--recursion-depth`-style naming while you'll still be able to access them in snake_case form—`recursion_depth`—in the Python code.

File: createtransfers/createtransfers.py [hunk 23]
```diff
@@ -0,0 +1,413 @@
...
+    """
+    for i in range(media_n):
+        new_media_fname = 'mediafile_{0}.{1}'.format(i, media_ext)
+        new_media_path = os.path.join(new_media_dir, new_media_fname)
+        shutil.copyfile(copy_media_path, new_media_path)
+    LOGGER.info("%s files created at %s", media_n, new_media_dir)
+
+
+def stat_file(file_path, files_n):
+    """ Check that the file exists to make sure we're not referencing objects
```
Comment (jrwdunham) [inline createtransfers/createtransfers.py]: What's with the `""" ` pattern here? I know this is nitpicky, but I would prefer docstrings to be consistent, at least within the module. Could we just consistently use `"""First words of docstring` instead of `""" First words of docstring` or `"""\nFirst words of docstring`?

**Review by jrwdunham — COMMENTED** [review]
This is looking good. I still have some stylistic comments.

File: createtransfers/createtransfers.py (hunk 23 above)
Comment (ross-spencer) [reply createtransfers/createtransfers.py]: I think this is just an innocent typo...

File: createtransfers/createtransfers.py (hunk 20 above)
Comment (ross-spencer) [reply createtransfers/createtransfers.py]: This change was made following pylint. Have you a configuration setup for the tool not to complain about the naming? Maybe my apt-repo has an older version?

Comment (jrwdunham) [reply createtransfers/createtransfers.py]: Ugh, that's annoying. You can add a comment to this module that will suppress the warning: `# pylint: disable=invalid-name`.

File: createtransfers/createtransfers.py (hunk 18 above)
Comment (ross-spencer) [reply createtransfers/createtransfers.py:190]: I hadn't thought about using isinstance like that. Way better.

File: createtransfers/createtransfers.py (hunk 7 above)
Comment (ross-spencer) [reply createtransfers/createtransfers.py]: Okay, I think I see... that is the path, but the file you previously had it pointed at `image_8.tif` isn't there:

https://github.com/artefactual/archivematica-sampledata/tree/master/TestTransfers/ManualNormalization/manualNormalization/preservation

It's here:

https://github.com/artefactual/archivematica-sampledata/blob/master/TestTransfers/manualNormalization/manualNormalization/preservation/image_8.tif

We have two folders with confusingly similar names! 

(I personally would like to overhaul this collection to make it a little more self-describing and give each folder a README.md. Not that it would have helped here.

**Comment by ross-spencer** [comment]
@jrwdunham that should fix those issues I hope. Thanks again for the CR on this. Lots learned as ever.

**Review by jrwdunham — APPROVED** [review]
